### PR TITLE
Ziemleszcz/uart

### PR DESCRIPTION
**Pull request by ziemleszcz**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 1]
```diff
@@ -145,48 +286,104 @@ void uart_thr(void *arg)
...
+	for (;;) {
+		do {
+			flags = __atomic_exchange_n(&uart.int_flags, 0, __ATOMIC_RELAXED);
+
+			if (flags & PARITY_ERR)
+				log_warn("parity error (%u)", uart.counters.parity);
+			if (flags & FRAME_ERR)
+				log_warn("frame error (%u)", uart.counters.frame);
+			if (flags & HW_OVERRUN_ERR)
+				log_warn("hw_overrun (%u)", uart.counters.hw_overrun);
```
Comment (nalajcie) [inline tty/imx6ull-uart/imx6ull-uart.c:368]: not sure if printing more chars (possibly on UART) on overflow is the best thing to do ;).

Side note: MISRA says we always need to use `{ }`

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 2]
```diff
@@ -301,20 +501,34 @@ void set_cflag(void* _uart, tcflag_t* cflag)
...
-static void signal_txready(void* _uart)
+
+static void signal_txready(void *_uart)
 {
-	uart_t* uartptr = (uart_t*) _uart;
+	/* disable TX interrupt to ensure that interrupt handler doesn't touch TX buffer */
+	*(uart.base + ucr1) &= ~UCR1_TRDYEN;
+
+	/* push as many characters as possible to HW FIFO at once */
+	while (!fifo_is_empty(uart.tty_common.tx_fifo)) {
```
Comment (nalajcie) [inline tty/imx6ull-uart/imx6ull-uart.c:511]: use `libtty_txready` and `libtty_getchar`?

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 3]
```diff
@@ -107,16 +247,17 @@ void uart_thr(void *arg)
 
 		switch (msg.type) {
 		case mtOpen:
-			// TODO: set PGID?
+			msg.o.io.err = uart_open(msg.i.openclose.flags);
```
Comment (nalajcie) [inline tty/imx6ull-uart/imx6ull-uart.c:250]: shame I needed to read kernel source code to understand how it works (but that's the problem of our API)

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 4]
```diff
@@ -77,21 +95,143 @@ typedef struct {
...
+{
+	/* TODO: set PGID? */
+	int id = flags_to_id(flags);
+
+	if (id <= 0)
+		return -EACCES;
+
+	mutexLock(uart.openclose_lock);
+
+	if (uart.busy & id) {
```
Comment (nalajcie) [inline tty/imx6ull-uart/imx6ull-uart.c:189]: ok, side effect is that I can't open UART twice (not that I need to, just it's possible on linux)

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 5]
```diff
@@ -77,21 +95,143 @@ typedef struct {
...
+
+/* interrupt flags */
+#define RX_DONE        (1 << 0)
+#define TX_DONE        (1 << 1)
+#define PARITY_ERR     (1 << 2)
+#define FRAME_ERR      (1 << 3)
+#define HW_OVERRUN_ERR (1 << 4)
+#define SW_OVERRUN_ERR (1 << 5)
+
+/* register flags */
```
Comment (nalajcie) [inline tty/imx6ull-uart/imx6ull-uart.c:138]: IMHO more readable than magic numbers but adding @HBuczynski to reviewers for kernel team to approve the style :P.

**Review by nalajcie — COMMENTED** [review]
Maybe we could use libtty external API (either change the semantics of existing functions or add some new non-{blocking,signaling} ones?)

File: tty/imx6ull-uart/imx6ull-uart.c (hunk 4 above)
Comment (ziemleszcz) [reply tty/imx6ull-uart/imx6ull-uart.c:189]: You mean two readers?

Comment (agkaminski) [reply tty/imx6ull-uart/imx6ull-uart.c:210]: Imho there is only a point in restricting readers. I see no problem with having multiple writers or reader and writer at the same time

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 6]
```diff
@@ -403,45 +655,59 @@ int main(int argc, char **argv)
 	if (portCreate(&port) != EOK)
 		return 2;
 
+	uart.rx_sw_fifo = malloc(sizeof(lf_fifo_t) + RX_SW_FIFO_SIZE * sizeof(uart.rx_sw_fifo->data[0]));
```
Comment (agkaminski) [inline tty/imx6ull-uart/imx6ull-uart.c:658]: All of malloc size arguments are const, this operation is always performed, imho let's avoid malloc if possible (it is possible here)

File: tty/imx6ull-uart/lf_fifo.h [hunk 7]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (agkaminski) [inline tty/imx6ull-uart/lf_fifo.h:1]: Maybe not now, but imho this should be in corelibs

File: tty/libtty/libtty.c [hunk 8]
```diff
@@ -170,21 +170,37 @@ ssize_t libtty_read_nonblock(libtty_common_t *tty, char *data, size_t size, unsi
...
-	if (fifo_freespace(tty->tx_fifo) >= TX_FIFO_NOTFULL_WATERMARK) {
-		if (wake_writer)
-			*wake_writer = 1;
-		condSignal(tty->tx_waitq);
+	if (wake_writer != NULL) {
+		*wake_writer = fifo_freespace(tty->tx_fifo) >= TX_FIFO_NOTFULL_WATERMARK;
+	}
+	else {
+		if (fifo_freespace(tty->tx_fifo) >= TX_FIFO_NOTFULL_WATERMARK)
+			condSignal(tty->tx_waitq);
```
Comment (agkaminski) [inline tty/libtty/libtty.c:184]: `libtty_wake_writer` fits here perfectly

**Review by agkaminski — COMMENTED** [review]
If you feel like it maybe adaptation to the new style (brackets everywhere)? But this is more a thing for separate PR. I see no point in adding a forest of debugging log_* macros, which are not even used anyway.

Register bits style is imho ok, it adheres to our new guidelines (avoiding enums used as int), but our convertion is rather to use bit number directly and comment it's use. `(1 << 2)` says not much, but so does `UCR3_RXDMUXSEL`. But `(1 << 2)` along with comment what we're doing (maybe even why we're doing it) is descriptive enough. You can define the bits however you like, but a comment is still advised and you still need to read the chip manual to perform changes, so TBH defines for bits seem to me like a waste of time and code lines.

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 9]
```diff
@@ -29,12 +32,71 @@
...
+#define COL_NORMAL "\033[0m"
+
+#define LOG_TAG "imx6ull-uart"
+
+/* clang-format off */
+#define log_debug(fmt, ...) do { log_printf(LOG_DEBUG, fmt "\n", ##__VA_ARGS__); } while (0)
+#define log_info(fmt, ...)  do { log_printf(LOG_INFO, fmt "\n", ##__VA_ARGS__); } while (0)
+#define log_warn(fmt, ...)  do { log_printf(LOG_WARNING, COL_YELLOW fmt COL_NORMAL "\n", ##__VA_ARGS__); } while (0)
+#define log_error(fmt, ...) do { log_printf(LOG_ERR, COL_RED fmt COL_NORMAL "\n", ##__VA_ARGS__); } while (0)
+/* clang-format on */
```
Comment (agkaminski) [inline tty/imx6ull-uart/imx6ull-uart.c:54]: Do we need this? I don't see these used... Only log_warn is used. Imho such logs should be for debug only, not enabled by default (so disabled under #if NDEBUG)

File: tty/imx6ull-uart/imx6ull-uart.c [hunk 10]
```diff
@@ -29,12 +32,71 @@
...
+
+#define READER 1
+#define WRITER 2
+
+#define MAX_READERS 255
```
Comment (agkaminski) [inline tty/imx6ull-uart/imx6ull-uart.c:59]: I thought we want only one reader to be allowed

File: tty/imx6ull-uart/imx6ull-uart.c (hunk 9 above)
Comment (ziemleszcz) [reply tty/imx6ull-uart/imx6ull-uart.c:54]: No, I can use `printf` directly if you don't like these defines. Currently I use `log_warn` to print UART errors which is useful for checking problems on DCU. These errors are not printed by default (option `-e`).

File: tty/imx6ull-uart/imx6ull-uart.c (hunk 10 above)
Comment (ziemleszcz) [reply tty/imx6ull-uart/imx6ull-uart.c:59]: I thought you have requested it. One reader is fine by me. Linux allows multiple readers.

**Comment by agkaminski** [comment]
Side note - don't ttys on NOMMU targets require change too?

File: tty/imx6ull-uart/imx6ull-uart.c (hunk 10 above)
Comment (agkaminski) [reply tty/imx6ull-uart/imx6ull-uart.c:59]: Ok, so misunderstanding, I thought we wanted to restrict to one reader only. I see no problem with allowing multiple readers on a driver side, it shouldn't really bother the driver who and how consumes the data.

File: tty/imx6ull-uart/imx6ull-uart.c (hunk 9 above)
Comment (agkaminski) [reply tty/imx6ull-uart/imx6ull-uart.c:54]: Ok, perhaps might be useful for the future debugging, developement

**Comment by ziemleszcz** [comment]
By NOMMU targets you mean multi drives, right? libtty API has not changed so they should be fine. I can replace `libtty_getchar` with `libtty_popchar` if you want.

**Comment by agkaminski** [comment]
> By NOMMU targets you mean multi drives, right? libtty API has not changed so they should be fine. I can replace `libtty_getchar` with `libtty_popchar` if you want.

It's ok, I'm just making sure these weren't overlooked.

**Pull request merged**